Delete web-based protocol handler page; fold relevant content into API method page (#35366)
mdn/content: files/en-us/web/api/navigator/registerprotocolhandler/index.md

- `scheme`

  - : A string containing the scheme for the protocol that the site wishes to handle.

    This may be a custom scheme, in which case the scheme's name:

    - Begins with `web+`
    - Contains at least one letter after the `web+` prefix
    - Contains only lowercase {{Glossary("ASCII")}} letters.

    Otherwise, the scheme must be one of the following:

    - `bitcoin`
    - `ftp`
    - `ftps`
    - `geo`
    - `im`
    - `irc`
    - `ircs`
    - `magnet`
    - `mailto`
    - `matrix`
    - `mms`
    - `news`
    - `nntp`
    - `openpgp4fpr`
    - `sftp`
    - `sip`
    - `sms`
    - `smsto`
    - `ssh`
    - `tel`
    - `urn`
    - `webcal`
    - `wtai`
    - `xmpp`

    <!-- This must match: https://html.spec.whatwg.org/multipage/system-state.html#safelisted-scheme -->

- `url`

  - : A string containing the URL of the handler.
    This URL must include `%s`, as a placeholder that will be replaced with the [escaped](/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) URL to be handled.

>     The handler URL must use the `https` scheme, and must be of the same {{glossary("origin")}} as the webpage that is attemtping to register the handler.

Reviewer: ```suggestion
    The handler URL must use the `https` scheme, and must be of the same {{glossary("origin")}} as the webpage that is attempting to register the handler.
```

    - The registered scheme (protocol) is invalid, such as a scheme the browser handles itself (`https:`, `about:`, etc.)
    - The handler URL's {{Glossary("origin")}} does not match the origin of the page calling this API.

>     - The browser requires that this function is called from a secure context.

Reviewer: I think this should be kept? Although it can also be inferred from the secure context header.

Author: I thought about this but it's implicitly true anyway: the template has to be HTTPS and it must be same-origin with the page. So if it throws with a non-HTTPS page, what caused the throw? It doesn't make sense to document it separately.

Interestingly most secure context APIs I've seen just don't exist outside a secure context, they don't throw. This one does seem to, though.

Author: > Interestingly most secure context APIs I've seen just don't exist outside a secure context, they don't throw. This one does seem to, though.

It turns out this is wrong (Firefox was silently upgrading http://example.org). With http://www.alwayshttp.com/, registerProtocolHandler is just not available, like other securecontext APIs. So the original bullet was wrong anyway.

Otherwise, the scheme must be one of the following:

- `bitcoin`
- `ftp`
- `ftps`
- `geo`
- `im`
- `irc`
- `ircs`
- `magnet`
- `mailto`
- `matrix`
- `mms`
- `news`
- `nntp`
- `openpgp4fpr`
- `sftp`
- `sip`
- `sms`
- `smsto`
- `ssh`
- `tel`
- `urn`
- `webcal`
- `wtai`
- `xmpp`

<!-- This must match: https://html.spec.whatwg.org/multipage/system-state.html#safelisted-scheme -->

## Examples

If your site is `burgers.example.com`, you can register a protocol handler for it to handle `web+burger:` links, like so:
In this example, a page served from `https://burgers.example.org/` contains a link like this:

```html
<a href="web+burger:cheeseburger">cheeseburger</a>
```

It registers the `web+burger` with code like this:

```js
navigator.registerProtocolHandler(
  "web+burger",
  "https://burgers.example.com/?burger=%s",
  "https://burgers.example.org/?burger=%s",
);
```

This creates a handler that lets `web+burger:` links send the user to your site, inserting the accessed burger URL into the `%s` placeholder.

This script must be run from the same origin as the handler URL (so any page at `https://burgers.example.com`), and the handler URL must be `http` or `https`.

> The user will be notified that your code asked to register the protocol handler, so that they can decide whether or not to allow it. See the screenshot below for an example on `google.co.uk`:

Reviewer: If FF asks the user (and they still do), I feel like we should still keep this note, and say it's FF-only? No strong opinions.

Author: I'm not exactly sure what Firefox's behavior is tbh. It does ask when you run the code (as well as when the user clicks a link) but it doesn't seem to make _that much_ difference what  you do: either way, you still get asked when the use clicks a link). So given that this is not AFAICT specced behavior, I thought it's better to be vague here than get into the details of what Firefox does.

    - The registered scheme (protocol) is invalid, such as a scheme the browser handles itself (`https:`, `about:`, etc.)
    - The handler URL's {{Glossary("origin")}} does not match the origin of the page calling this API.
    - The handler's URL's scheme is not `https`.

- `SyntaxError` {{domxref("DOMException")}}
  - : The `%s` placeholder is missing from the handler URL.

## Examples

### Registering a handler for the mailto protocol

It's fairly common for web pages to link to resources using non-`https` protocols. An example is the `mailto:` protocol. Web authors can use a `mailto` link when they want to provide a convenient way for users to send an email directly from the webpage:

```html
<a href="mailto:webmaster@example.com">Web Master</a>
```

When the link is activated, the browser should launch the default desktop application for handling email. You can think of this as a _desktop-based_ protocol handler.

> Web-based protocol handlers allow web-based applications to participate in the process too. A webpage can register to handle `mailto` links with code like this:

Reviewer: ```suggestion
Web-based protocol handlers allow web-based applications to participate in the process too. A mail web app at `mail.example.org` can register to handle `mailto` links with code like this:
```

```js

> navigator.registerProtocolHandler("mailto", "https://mail.example.org/?to=%s");

Reviewer: Worth mentioning that the page must be same-origin as this `mailto` handler, so the handler URL may need to do a second redirect if it wants to go to a mail site like gmail

Author: Isn't this already documented here: https://pr35366.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/Navigator/registerProtocolHandler#securityerror ?

Reviewer: Yes, I meant that they wouldn't misinterpret this example as you could directly open gmail using `mailto:`.

```

> After this, when the user visits a page containing a `mailto` link, the browser will (after possibly asking the browser for confirmation) navigate to `https://mail.example.org/?to=mailto:webmaster@example.com`. This page could parse the URL parameter to extract the address, and use this to initialize an email.

Reviewer: ```suggestion
After this, when the user clicks a `mailto` link on any website, the browser will (after possibly asking the browser for confirmation) navigate to `https://mail.example.org/?to=mailto:webmaster@example.com`. This page could parse the URL parameter to extract the address, and use this to initialize an email.
```
